### PR TITLE
User/pinkb/ambient occlusion

### PR DESCRIPTION
**Pull request by pinkwerks**
## Overview

Adds an "ambient occlusion" component that modifies a mesh to include ambient occlusion data for use by the Standard Shader

image

image

## Motivation

Ambient occlusion adds realism to a scene by providing important clues to viewers about an object's relationship to other objects. However, as a post-process it is expensive and not well suited for VR and low-end platforms. Baking global illumination to textures can cumbersome and relies on sane UVs (when not using lightmaps). 

As an alternative, we bake this occlusion information into the mesh vertices, which is evaluated per vertex and passed to fragment shader as part of rasterization.

## Notes
- Additional updates to `StandardShaderGUI.cs` to support material conversions that use GLTF material semantics
- The fidelity of the AO solution is very dependent on the vertex count and normals of the mesh.

### CODE REVIEW

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+            UnityEngine.Debug.Log($"{nameof(GatherSamples)} elapsed-ms={watch.ElapsedMilliseconds} vertex-count={SourceMesh.vertexCount} rays-per-vertex={SamplesPerVertex}.");
+        }
+
+        private void SaveOriginalColors()
+        {
+            UnityEngine.Debug.Log(nameof(SaveOriginalColors));
+            if (SourceMesh.colors.Length == SourceMesh.vertexCount)
+            {
+                _originalHasColors = true;
+                _originalColors = new Color[SourceMesh.vertexCount];
```
Comment (Cameron-Micka) [inline com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs:362]:
```suggestion
                _originalColors = SourceMesh.colors;
```

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+                Array.Copy(SourceMesh.colors, _originalColors, SourceMesh.vertexCount);
+                SourceMesh.GetUVs(kUvChannel, _originalUv4);
+            }
+        }
+
+        private void RestoreColors()
+        {
+            // XXX see if we can reuse old data without re-sample
+            UnityEngine.Debug.Log(nameof(RestoreColors));
```
Comment (Cameron-Micka) [inline com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs:371]: nit: do you need this?

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+            if (_originalColors == null)
+            {
+                SaveOriginalColors();
+            }
+
+            var colors = new Color[_originalColors.Length];
+
+            if (_originalHasColors)
+            {
+                for (int i = 0; i < _originalColors.Length; i++)
```
Comment (Cameron-Micka) [inline com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs:385]: Do you need this loop or can you just assign _originalColors to SourceMesh.colors?

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Shaders/GraphicsToolsStandardProgram.hlsl [hunk 4]
```diff
@@ -762,26 +768,31 @@ half4 PixelStage(Varyings input, bool facing : SV_IsFrontFace) : SV_Target
...
+#if defined(_SPHERICAL_HARMONICS)
     half3 bakedGI = input.ambient;
 #else
-    half3 bakedGI = glstate_lightmodel_ambient.rgb + half3(0.25h, 0.25h, 0.25h);
+    half3 bakedGI = glstate_lightmodel_ambient.rgb * occlusion + half3(0.25h, 0.25h, 0.25h);
 #endif
 
     // Indirect lighting.
-    output.rgb = GTGlobalIllumination(brdfData, bakedGI, occlusion, worldNormal, worldViewDir);
+
```
Comment (Cameron-Micka) [inline com.microsoft.mrtk.graphicstools.unity/Runtime/Shaders/GraphicsToolsStandardProgram.hlsl:792]: nit awkward space
```suggestion
```

**Review by Cameron-Micka — COMMENTED** [review]
This is super cool and useful. Thank you for the contribution! Added some minor comments below.

I also tried it out locally and saw AO generated as expected (circled in blue) but also saw some vertices appear to get brighter (circled in green). Is this expected? 
image

**Comment by pinkwerks** [comment]
> This is super cool and useful. Thank you for the contribution! Added some minor comments below.
> 
> I also tried it out locally and saw AO generated as expected (circled in blue) but also saw some vertices appear to get brighter (circled in green). Is this expected? image

Thanks for the review and comments! The "brighter" vertex was un-occluded due to ray casts not considering back faces. This has been fixed.

File: com.microsoft.mrtk.graphicstools.unity/Editor/Experimental/AmbientOcclusion/AmbientOcclusionEditor.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
+using UnityEditor;
```
[the remark this answers is not in the document]
Comment (pinkwerks) [reply com.microsoft.mrtk.graphicstools.unity/Editor/Experimental/AmbientOcclusion/AmbientOcclusionEditor.cs:4]: Fixed.

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+
+        private void Reset()
+        {
+            UnityEngine.Debug.Log("Reset");
+            SaveOriginalColors();
+        }
+
+        private void Awake()
+        {
+            UnityEngine.Debug.Log("Awake");
```
[the remark this answers is not in the document]
Comment (pinkwerks) [reply com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs]: Nope, fixed.

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+
+        private void OnEnable()
+        {
+            UnityEngine.Debug.Log("OnEnable");
+            SaveOriginalColors();
+        }
+
+        private void OnDisable()
+        {
+            UnityEngine.Debug.Log("OnDisable");
```
[the remark this answers is not in the document]
Comment (pinkwerks) [reply com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs]: Fixed.

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+        }
+
+        private void Start()
+        {
+            UnityEngine.Debug.Log("Start");
```
[the remark this answers is not in the document]
Comment (pinkwerks) [reply com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs]: Fixed.

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+        }
+
+        private void OnValidate()
+        {
+            UnityEngine.Debug.Log("OnValidate");
+
+            SamplesPerVertex = Mathf.Max(1, SamplesPerVertex);
```
[the remark this answers is not in the document]
Comment (pinkwerks) [reply com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs]: Great tip! Fixed.

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+        /// </summary>
+        /// <param name="normalizedReferenceNormal">Must be noramlized</param>
+        /// <returns></returns>
+        private Vector3 RandomSampleAboveHemisphere(Vector3 normalizedReferenceNormal)
+        {
+            float angle;
+            Vector3 randomAxis;
+            while (true)
+            {
+                UnityEngine.Random.rotationUniform.ToAngleAxis(out angle, out randomAxis);
```
[the remark this answers is not in the document]
Comment (pinkwerks) [reply com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs:226]: I tested both methods and discovered random rotation to be slightly more uniform than on unity sphere given 10,000 samples. Keeping the very slightly slower method given the better distribution.

Random-unitsphere-vs-uniformrotation

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+            if (_bentNormalsAo == null || _bentNormalsAo.Length != SourceMesh.vertexCount)
+            {
+                _bentNormalsAo = new Vector4[SourceMesh.vertexCount];
+            }
+
+            var watch = Stopwatch.StartNew();
+
+            UnityEngine.Random.InitState(Seed);
+
+            _vertexes = new Vector3[SourceMesh.vertexCount];
```
[the remark this answers is not in the document]
Comment (pinkwerks) [reply com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs]: Fixed.

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs (hunk 1 above)
Comment (pinkwerks) [reply com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs]: Fixed.

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs [hunk 12]
```diff
@@ -0,0 +1,394 @@
...
+            Array.Copy(SourceMesh.vertices, _vertexes, SourceMesh.vertexCount);
+
+            _normals = new Vector3[SourceMesh.vertexCount];
```
[the remark this answers is not in the document]
Comment (pinkwerks) [reply com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs]: Fixed

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs (hunk 2 above)
Comment (pinkwerks) [reply com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs]: Nope, fixed.

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs (hunk 3 above)
Comment (pinkwerks) [reply com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs]: Fixed.

File: com.microsoft.mrtk.graphicstools.unity/Runtime/Shaders/GraphicsToolsStandardProgram.hlsl (hunk 4 above)
Comment (pinkwerks) [reply com.microsoft.mrtk.graphicstools.unity/Runtime/Shaders/GraphicsToolsStandardProgram.hlsl:792]: Fixed

**Comment by Cameron-Micka** [comment]
Just checking in to see if this is ready for a re-review?

**Comment by pinkwerks** [comment]
I'm going to refactor this into a standalone window, and not a component. Will re-open a fresh PR when that is ready.

**Pull request closed without merging**